content/post/2019-06-18.md

```diff
@@ -0,0 +1,256 @@
+
+```
+nodedef>name VARCHAR,label VARCHAR
+s1,Site number 1
+s2,Site number 2
+s3,Site number 3
+edgedef>node1 VARCHAR,node2 VARCHAR
+s1,s2
+s2,s3
+s3,s2
+s3,s1
+```
+
+The steps to extract relationships from a CSV file and generate a file in GDF format are:
+
+1. Define the parameters for the extraction of related data
+2. Define the data structures for the transformation
+3. Import the data from a file in CSV format and store the data in the structures
+4. Sort data according to connections
+5. Prepare the data for the GDF format
+6. Generate the file in GDF format
+
+## Define the parameters for the extraction of related data
+
+A set of parameters are defined so this converter can be adapted to different cases. We must specify what information will be taken from the file in CSV format, where we will leave the result and the graph type.
+
+```r
+library("readr")
+library("hash")
+library("stringr")
+
+source=3
+target=9
+source_attribs= c(5,6,11,12,13)
+null_attrib=c(NA,NA,NA,NA,NA)
+name_attribs=c("app","location","hastag","lang","create at")
+name_file_csv<-"rstudio_RTs.csv"
+name_file_gdf<-"rstudio_RTs.gdf"
+directed <-TRUE
+```
+
+The sample CSV file can be found [here](/post/2019-06-18_files/rstudio_RTs.csv).
+
+## Define the data structures for the transformation
+
+In order to store the data, dynamic structures are needed to allow data to be added as they appear. The hash tables were chosen because they are the most appropriate for this case.
+Hash tables will be used to store nodes and connections.
+
+```r
+hash_nodes <- hash()
+hash_links <- hash()
+hash_links_in <- hash()
+hash_links_out <- hash()
+hash_connections <- hash()
+hash_connections_attrib<- hash()
+num_attribs<-length(source_attribs)
+```
+
+## Import the data from a file in CSV format and store the data in the structures
+
+Import data reading the CSV file and run it row by row to store the nodes and connections in the hash tables.
+
+Related entities can appear multiple times, as a source or as a target. When an entity appears for the first time, it is stored in the hash_nodes table. Attributes are associated to source entities and null attributes to target entities. It is a criterion that assumes this algorithm, but there could be others. If an entity appears first as a target as source, the null attributes are replaced by theirs as source.
+
+For each entity, the number of total links (hash_links), the number of inbound links (hash_links_in) and the number of outbound links (hash_links) are counted. This is done to allow ordering the nodes from greater to lesser degree when generating the file in GDF format.
+
+For each origin-target entity pair, the number of times that relation appears (hash_connections) and the attributes (hash_connections_attrib) are stored. In the first case, we get the weight of the relationship and the second we get the associated attributes.
```

Comment:
Creo que falta un THE en la siguiente frase: For each origin... of times that THE relation appears...

Y un "IN", en la siguiente frase a compañando a "second": In the first case... and, IN the second,...
Yo de hecho, también pondría "in the second" entre comas.

Reply:
ok